arkouda/alignment.py

```diff
@@ -425,43 +425,62 @@ def search_intervals(vals, intervals):
         valid = functools.reduce(lambda x, y: x & y, [e > s + 1 for e, s in zip(ends, starts)])
         if valid.sum() > 0:
-            # pranges is all the indices in sorted array that fall between a lower and an uppper bound
+            # pranges is all the indices in sorted array that fall between a lower and
+            # an uppper bound for each dimension
             segs, pranges = zip(*[gen_ranges(s[valid] + 1, e[valid]) for s, e in zip(starts, ends)])
             # matches are the indices of those items in the original array
             matches = [pm[pr] for pm, pr in zip(perm, pranges)]
-            # integer indices of each interval containing a hit
+            # integer indices of each one-dimensional interval containing a hit
             hitidx = arange(valid.size)[valid]
-            # broadcast interval index out to matches
+            # broadcast 1-d interval index out to matches
             matchintervalidx = [broadcast(s, hitidx, m.size) for s, m in zip(segs, matches)]
             # make sure not to include any of the bounds themselves
             validmatch = [(m >= lowsize) & (m < boundary) for m in matches]
-            # indices of unique values found (translated from concat keys)
+            # indices of values found (translated from concat keys) in 1-d intervals
             uvalidx = [m[vm] - lowsize for m, vm in zip(matches, validmatch)]
-            # set index of containing interval for uvals that were found
-            for ci, uv, mi, vm in zip(containinginterval, uvalidx, matchintervalidx, validmatch):
-                ci[uv] = mi[vm]
-            return functools.reduce(lambda x, y: where(x == y, x, -1), containinginterval)
+            # now go from 1-d to full dimensionality
+            # for each interval, intersect the hits from its 1-d projections
+            # do this by concatenating all the projections and grouping on the id of the hit
+            # and the interval and looking for hits that cover all dimensions
+            alluvalidx = concatenate(uvalidx, ordered=False)
+            allmatchintervalidx = concatenate(
+                [mi[vm] for mi, vm in zip(matchintervalidx, validmatch)], ordered=False
+            )
+            byvalinterval = GroupBy([alluvalidx, allmatchintervalidx])
+            # a true hit happens when a value is contained in all of an interval's 1-d projections
+            isahit = byvalinterval.count()[1] == len(low)
+            # indices of the true hits and their containing intervals
+            valhits, intervalhits = [x[isahit] for x in byvalinterval.unique_keys]
+            # a value might be found in more than one interval, so we need to break ties
+            byval = GroupBy(valhits)
+            hitvalidx, tmpidx = byval.argmin(tiebreak[intervalhits])
+            winninginterval = intervalhits[tmpidx]
+            # set index of best containing interval for values that were found
+            containinginterval[hitvalidx] = winninginterval
```

Comment:
this is a def nitpicky, but is there a reason we don't do this on one line?
```python
containinginterval[hitvalidx] = intervalhits[tmpidx]
```
This feels more symmetrical to the single pdarray case